OutOfSchool/OutOfSchool.WebApi/Services/Database/WorkshopService.cs

```diff
@@ -266,22 +269,100 @@ public async Task<SearchResult<T>> GetByProviderId<T>(Guid id, ExcludeIdFilter f
             filter.ExcludedId == null
                 ? (x.ProviderId == id)
                 : (x.ProviderId == id && x.Id != filter.ExcludedId)).ConfigureAwait(false);
-        var workshops = await workshopRepository.Get(
+
+        var workshops = workshopRepository.Get(
                 skip: filter.From,
                 take: filter.Size,
                 includeProperties: includingPropertiesForMappingDtoModel,
                 where: x => filter.ExcludedId == null
                     ? (x.ProviderId == id)
-                    : (x.ProviderId == id && x.Id != filter.ExcludedId))
-            .ToListAsync()
-            .ConfigureAwait(false);
-
-        logger.LogInformation(!workshops.Any()
+                    : (x.ProviderId == id && x.Id != filter.ExcludedId));
+
+        var chatrooms = roomRepository.Get(
+            includeProperties: $"{nameof(ChatRoomWorkshop.ChatMessages)}",
+            where: x => x.ChatMessages.Where(x => x.ReadDateTime == null && !x.SenderRoleIsProvider).Any());
+
+        var query = from w in workshops
+                    join c in chatrooms
+                    on w.Id equals c.WorkshopId
+                    into g
+                    from cg in g.DefaultIfEmpty()
+                    select new
+                    {
+                        w.Id,
+                        w.Title,
+                        w.Phone,
```

Comment:
(Performance and readability) 
1. Excessive fields
You are asking `query` to return lot of fields that are not used anywhere (I won'tname all, but `Phone` and `DateTimeRanges` aren't used). 
This is bad from **performance** perspective - because they are transfered from db to the app.
This is bad from **readability** and **code maintainability** perspective - your anonymous object creation takes 30 lines of your code

2. Memory pressure, excessive objects and GroupJoin.
Theory: `GroupJoin` (see [complex query operators](https://learn.microsoft.com/en-us/ef/core/querying/complex-query-operators#groupjoin)) as well as your `into` operator doesn't execute on db side (`EF Core doesn't translate GroupJoin` - from mentioned article). So initially you obtain all records from db and then on C# side EF does the group join. 
In your case: for the each grouping (done on C# side) you have intermediate anonymous object (`select new { w.Id ....`), that stores necessary data. And immediately after that (smaller) part of that data is mapped into `WorkshopProviderViewCard`, and then all that anon objects are no longer used.

This is bad from **performance** perspective - you have to copy data twice (to anon object and then to `WorkshopProviderViewCard`).
This is bad from **memory usage** perspective - you are bothering GC with ephemeral allocations (all that anon objects).
This is bad from **readability** and **code maintainability** perspective - extra 25 lines of code are needed to copy data second time, and it's hard to see in one place if previously mentioned excessive fields could be removed.